### PR TITLE
Updates for visualization CI

### PR DESCRIPTION
**Pull request by jorgensd**
- Add IPython to CI
- Bump `pyvista` and `pyvistaqt` version
- Use `python3 -m pip` over `pip3`
- Change matplotlib backend for TNT demo (to avoid CI issues)
- Remove `pythreejs` and `ipygany` from notebooks, as `trame` is now preferred

### CODE REVIEW

**Comment by garth-wells** [comment]
@jorgensd does this fix the CI?

**Comment by jorgensd** [comment]
> @jorgensd does this fix the CI?

The CI has a strange issue with the tnt demo, that doesn’t use pyvista https://github.com/FEniCS/dolfinx/actions/runs/4508930557/jobs/7940462505
ref
https://docs.fenicsproject.org/dolfinx/v0.6.0/python/demos/demo_tnt-elements.html

**Comment by garth-wells** [comment]
But it does use Matplotlib.

**Comment by jorgensd** [comment]
> But it does use Matplotlib.

Running the code locally, I get issues at import, just running
```python
import numpy as np

from dolfinx import fem, mesh
import basix
import basix.ufl_wrapper
```
throws
```bash
root@dokken-XPS-15-9560:~/shared/dolfinx/python/demo# mpirun -n 2 python3 -m pytest -xsv test.py::test_demos[path12-demo_tnt-elements.py]
============================= test session starts ==============================
platform linux -- Python 3.10.4, pytest-7.1.2, pluggy-1.0.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /root/shared/dolfinx/python/demo, configfile: pytest.ini
plugins: forked-1.4.0, xdist-2.5.0
collecting ... ============================= test session starts ==============================
platform linux -- Python 3.10.4, pytest-7.1.2, pluggy-1.0.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /root/shared/dolfinx/python/demo, configfile: pytest.ini
plugins: forked-1.4.0, xdist-2.5.0
collecting ... collected 1 item

test.py::test_demos[path12-demo_tnt-elements.py] collected 1 item

test.py::test_demos[path12-demo_tnt-elements.py] [cli_0]: write_line error; fd=6 buf=:cmd=init pmi_version=1 pmi_subversion=1
:
system msg for write_line failure : Bad file descriptor
[cli_0]: Unable to write to PMI_fd
[cli_0]: write_line error; fd=6 buf=:cmd=get_appnum
:
system msg for write_line failure : Bad file descriptor
[cli_1]: write_line error; fd=8 buf=:cmd=init pmi_version=1 pmi_subversion=1
:
system msg for write_line failure : Bad file descriptor
[cli_1]: Unable to write to PMI_fd
[cli_1]: write_line error; fd=8 buf=:cmd=get_appnum
:
system msg for write_line failure : Bad file descriptor
Abort(1691407): Fatal error in internal_Init_thread: Other MPI error, error stack:
internal_Init_thread(60): MPI_Init_thread(argc=0x7f181f2e8690, argv=0x7f181f2e8688, required=1, provided=0x7fffb5f0a998) failed
MPII_Init_thread(209)...: 
MPID_Init(359)..........: 
MPIR_pmi_init(110)......: PMI_Get_appnum returned -1
Abort(471453455): Fatal error in internal_Init_thread: Other MPI error, error stack:
internal_Init_thread(60): MPI_Init_thread(argc=0x7fe3f7cc6690, argv=0x7fe3f7cc6688, required=1, provided=0x7fffad3f9758) failed
MPII_Init_thread(209)...: 
MPID_Init(359)..........: 
MPIR_pmi_init(110)......: PMI_Get_appnum returned -1
FAILED

=================================== FAILURES ===================================
___________________ test_demos[path12-demo_tnt-elements.py] ____________________

path = PosixPath('/root/shared/dolfinx/python/demo')
name = 'demo_tnt-elements.py'

    @pytest.mark.serial
    @pytest.mark.parametrize("path,name", demos)
    def test_demos(path, name):
>       ret = subprocess.run([sys.executable, name], cwd=str(path), check=True)

test.py:26: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

input = None, capture_output = False, timeout = None, check = True
popenargs = (['/usr/bin/python3', 'demo_tnt-elements.py'],)
kwargs = {'cwd': '/root/shared/dolfinx/python/demo'}
process = <Popen: returncode: 15 args: ['/usr/bin/python3', 'demo_tnt-elements.py']>
stdout = None, stderr = None, retcode = 15

    def run(*popenargs,
            input=None, capture_output=False, timeout=None, check=False, **kwargs):
        """Run command with arguments and return a CompletedProcess instance.
    
        The returned instance will have attributes args, returncode, stdout and
        stderr. By default, stdout and stderr are not captured, and those attributes
        will be None. Pass stdout=PIPE and/or stderr=PIPE in order to capture them.
    
        If check is True and the exit code was non-zero, it raises a
        CalledProcessError. The CalledProcessError object will have the return code
        in the returncode attribute, and output & stderr attributes if those streams
        were captured.
    
        If timeout is given, and the process takes too long, a TimeoutExpired
        exception will be raised.
    
        There is an optional argument "input", allowing you to
        pass bytes or a string to the subprocess's stdin.  If you use this argument
        you may not also use the Popen constructor's "stdin" argument, as
        it will be used internally.
    
        By default, all communication is in bytes, and therefore any "input" should
        be bytes, and the stdout and stderr will be bytes. If in text mode, any
        "input" should be a string, and stdout and stderr will be strings decoded
        according to locale encoding, or by "encoding" if set. Text mode is
        triggered by setting any of text, encoding, errors or universal_newlines.
    
        The other arguments are the same as for the Popen constructor.
        """
        if input is not None:
            if kwargs.get('stdin') is not None:
                raise ValueError('stdin and input arguments may not both be used.')
            kwargs['stdin'] = PIPE
    
        if capture_output:
            if kwargs.get('stdout') is not None or kwargs.get('stderr') is not None:
                raise ValueError('stdout and stderr arguments may not be used '
                                 'with capture_output.')
            kwargs['stdout'] = PIPE
            kwargs['stderr'] = PIPE
    
        with Popen(*popenargs, **kwargs) as process:
            try:
                stdout, stderr = process.communicate(input, timeout=timeout)
            except TimeoutExpired as exc:
                process.kill()
                if _mswindows:
                    # Windows accumulates the output in a single blocking
                    # read() call run on child threads, with the timeout
                    # being done in a join() on those threads.  communicate()
                    # _after_ kill() is required to collect that and add it
                    # to the exception.
                    exc.stdout, exc.stderr = process.communicate()
                else:
                    # POSIX _communicate already populated the output so
                    # far into the TimeoutExpired exception.
                    process.wait()
                raise
            except:  # Including KeyboardInterrupt, communicate handled that.
                process.kill()
                # We don't call process.wait() as .__exit__ does that for us.
                raise
            retcode = process.poll()
            if check and retcode:
>               raise CalledProcessError(retcode, process.args,
                                         output=stdout, stderr=stderr)
E               subprocess.CalledProcessError: Command '['/usr/bin/python3', 'demo_tnt-elements.py']' returned non-zero exit status 15.

/usr/lib/python3.10/subprocess.py:524: CalledProcessError
=========================== short test summary info ============================
FAILED test.py::test_demos[path12-demo_tnt-elements.py] - subprocess.CalledPr...
!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!
============================== 1 failed in 1.73s ===============================
FAILED

=================================== FAILURES ===================================
___________________ test_demos[path12-demo_tnt-elements.py] ____________________

path = PosixPath('/root/shared/dolfinx/python/demo')
name = 'demo_tnt-elements.py'

    @pytest.mark.serial
    @pytest.mark.parametrize("path,name", demos)
    def test_demos(path, name):
>       ret = subprocess.run([sys.executable, name], cwd=str(path), check=True)

test.py:26: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

input = None, capture_output = False, timeout = None, check = True
popenargs = (['/usr/bin/python3', 'demo_tnt-elements.py'],)
kwargs = {'cwd': '/root/shared/dolfinx/python/demo'}
process = <Popen: returncode: 15 args: ['/usr/bin/python3', 'demo_tnt-elements.py']>
stdout = None, stderr = None, retcode = 15

    def run(*popenargs,
            input=None, capture_output=False, timeout=None, check=False, **kwargs):
        """Run command with arguments and return a CompletedProcess instance.
    
        The returned instance will have attributes args, returncode, stdout and
        stderr. By default, stdout and stderr are not captured, and those attributes
        will be None. Pass stdout=PIPE and/or stderr=PIPE in order to capture them.
    
        If check is True and the exit code was non-zero, it raises a
        CalledProcessError. The CalledProcessError object will have the return code
        in the returncode attribute, and output & stderr attributes if those streams
        were captured.
    
        If timeout is given, and the process takes too long, a TimeoutExpired
        exception will be raised.
    
        There is an optional argument "input", allowing you to
        pass bytes or a string to the subprocess's stdin.  If you use this argument
        you may not also use the Popen constructor's "stdin" argument, as
        it will be used internally.
    
        By default, all communication is in bytes, and therefore any "input" should
        be bytes, and the stdout and stderr will be bytes. If in text mode, any
        "input" should be a string, and stdout and stderr will be strings decoded
        according to locale encoding, or by "encoding" if set. Text mode is
        triggered by setting any of text, encoding, errors or universal_newlines.
    
        The other arguments are the same as for the Popen constructor.
        """
        if input is not None:
            if kwargs.get('stdin') is not None:
                raise ValueError('stdin and input arguments may not both be used.')
            kwargs['stdin'] = PIPE
    
        if capture_output:
            if kwargs.get('stdout') is not None or kwargs.get('stderr') is not None:
                raise ValueError('stdout and stderr arguments may not be used '
                                 'with capture_output.')
            kwargs['stdout'] = PIPE
            kwargs['stderr'] = PIPE
    
        with Popen(*popenargs, **kwargs) as process:
            try:
                stdout, stderr = process.communicate(input, timeout=timeout)
            except TimeoutExpired as exc:
                process.kill()
                if _mswindows:
                    # Windows accumulates the output in a single blocking
                    # read() call run on child threads, with the timeout
                    # being done in a join() on those threads.  communicate()
                    # _after_ kill() is required to collect that and add it
                    # to the exception.
                    exc.stdout, exc.stderr = process.communicate()
                else:
                    # POSIX _communicate already populated the output so
                    # far into the TimeoutExpired exception.
                    process.wait()
                raise
            except:  # Including KeyboardInterrupt, communicate handled that.
                process.kill()
                # We don't call process.wait() as .__exit__ does that for us.
                raise
            retcode = process.poll()
            if check and retcode:
>               raise CalledProcessError(retcode, process.args,
                                         output=stdout, stderr=stderr)
E               subprocess.CalledProcessError: Command '['/usr/bin/python3', 'demo_tnt-elements.py']' returned non-zero exit status 15.

/usr/lib/python3.10/subprocess.py:524: CalledProcessError
=========================== short test summary info ============================
FAILED test.py::test_demos[path12-demo_tnt-elements.py] - subprocess.CalledPr...
!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!
============================== 1 failed in 1.75s ===============================
```
It only happens when running the code through pytest, not as
```bash
mpirun -n 2 python3 demo_tnt-elements.py 
```

**Comment by jorgensd** [comment]
Sigh, seem like a random fluke that it passed

**Comment by jorgensd** [comment]
Now passing again @garth-wells: https://github.com/FEniCS/dolfinx/actions/runs/4540817530/jobs/8002207371 
I changed the backend of matplotlib to not use QT to avoid any issues with pyvista using xvfb

File: .github/workflows/pyvista.yml [hunk 1]
```diff
@@ -62,7 +62,9 @@ jobs:
...
       - name: Run pyvista demos (Python, serial)
-        run: python3 -m pytest -v -n=2 -m serial --durations=10 python/demo/test.py
+        run:
+          python3 -m pytest -v -n=2 -m serial --durations=10 python/demo/test.py
 
-      - name: Run pyivsta demos (Python, MPI (np=2))
-        run: python3 -m pytest -v -m mpi --num-proc=2 python/demo/test.py
+      - name: Run pyvista demos (Python, MPI (np=2))
+        run:
+          python3 -m pytest -v -m mpi --num-proc=2 python/demo/test.py
```
Comment (jorgensd) [inline .github/workflows/pyvista.yml:70]:
```suggestion
      - name: Run pyvista demos (Python, MPI (np=2))
        run: python3 -m pytest -v -m mpi --num-proc=2 python/demo/test.py
```

File: python/demo/demo_tnt-elements.py [hunk 2]
```diff
@@ -32,6 +33,8 @@
 
 from mpi4py import MPI
 
+matplotlib.use('agg')
```
Comment (garth-wells) [inline python/demo/demo_tnt-elements.py:36]: @jorgensd could you check that this doesn't affect the Python docs? The docs run this demo to created plots.

Comment (jorgensd) [reply python/demo/demo_tnt-elements.py:36]: Runs fine locally:

image

image

**Pull request merged**